library/Icingadb/Common/Macros.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+/* Icinga DB Web | (c) 2021 Icinga GmbH | GPLv2 */
+
+namespace Icinga\Module\Icingadb\Common;
+
+use Icinga\Application\Logger;
+use Icinga\Module\Icingadb\Model\Host;
+use Icinga\Module\Icingadb\Model\Service;
+
+trait Macros
+{
+    /**
+     * Known icinga macros
+     *
+     * @var array
+     */
+    protected $icingaMacros = [
+        'host.name'     => 'name',
+        'host.address'  => 'address',
+        'host.address6' => 'address6',
+        'service.name'  => 'name'
+    ];
+
+    /**
+     * Get the given string with macros being resolved
+     *
+     * @param string       $input  The string in which to look for macros
+     * @param Host|Service $object The host or service used to resolve the macros
+     *
+     * @return string
+     */
+    public function expandMacros($input, $object)
+    {
+        if (preg_match_all('@\$([^\$\s]+)\$@', $input, $matches)) {
+            foreach ($matches[1] as $key => $value) {
+                $newValue = $this->resolveMacro($value, $object);
+                if ($newValue !== $value) {
+                    $input = str_replace($matches[0][$key], $newValue, $input);
+                }
+            }
+        }
+
+        return $input;
+    }
+
+    /**
+     * Resolve a macro based on the given object
+     *
+     * @param string       $macro  The macro to resolve
+     * @param Host|Service $object The host or service used to resolve the macros
+     *
+     * @return string
+     */
+    public function resolveMacro($macro, $object)
```

Comment:
It's still not strict enough. `service.name` will currently be evaluated also if `$object` is a host. Though services are not joined for hosts and even then, it'd be a list of services not just a single one. `service.name` is the **only** macro that is supported with this prefix and can only be evaluated on services to differentiate it from `host.name` (`$service->host` exists and is then targeted)